### ISSUE TITLE
[Question] Clarification regarding file upload progress number

### ISSUE THREAD

**Issue by nik32** [ryanto/next-s3-upload#95]
First of all, Great library with to the point docs!!!! Thanks guys its really has been of great help!!!

<img width="902" alt="image" src="https://user-images.githubusercontent.com/51018572/190033790-434632d7-0c72-4e2f-9be6-51b86d6d93be.png">
@ryanto In the docs it's stated that the progress number will be 'continuously' updated as file's being uploaded.

But when I am uploading a small file (lets say <10mb) - the number doesn't upload [it just directly reaches from 0-100]. In case of bigger files, it does update - but that is also in big chunks [like from 0 - 40, to then directly 100].

Is this the intended behaviour or am I missing something??

**Comment by ryanto** [comment]
Hi @nik32 sorry I missed this question!

It should update continuously, but I think for smaller files the upload might happen in one part so the only update is 0 to 100.

If you slow your internet down in chrome dev tools (to slow 3g) does it still only update once? it would be worth trying a few scenarios with small files and slow internet to see how the progress is reported.

The AWS SDK has the ability to change part size, so we could allow that to be configured when calling `uploadToS3`. However, messing with the defaults can sometimes create more headaches so I'd like to avoid that if at all possible.

**Comment by nik32** [comment]
Hi @ryanto, thanks for your replay!!! So I tried for 2 files [one of 900KB and the other of 5.3MB] with slow 3G, and the result was same [it was going directly from 0 to 100]!!

**Comment by ryanto** [comment]
Ok interesting, that helps! I'll start messing with the part size and see if I can get better progress updates from the aws-sdk. If that goes well I think we can provide some sort of fix/control over how often you get updates.

It might take me a week or two to get started on this.

**Comment by nik32** [comment]
> However, messing with the defaults can sometimes create more headaches so I'd like to avoid that if at all possible.

Thanks @ryanto!!! Do try it, but if it ends up creating more problems at other places, leave it.

**Comment by ryanto** [comment]
Just looked into this, and the aws-sdk has a minimum part size of 5mb, which unfortunately means that you'll only get progress updates every 5mb.

There's an issue here about using the XHR adapter when doing the upload to get faster feedback: https://github.com/aws/aws-sdk-js-v3/issues/3101

**Comment by nik32** [comment]
Hi @ryanto!! sorry for the late replay and thank you for your efforts. I went into this [aws/aws-sdk-js-v3#3101]( https://github.com/aws/aws-sdk-js-v3/issues/3101) and in that especially this one [XHRHttpHandller](https://www.npmjs.com/package/@aws-sdk/xhr-http-handler). So do I need to implement this my our project code or would it be better if you do it inside the library code and provide us some property to opt into it???

Also does this completely replaces `fetch` as our http request handler [coz in the docs they use the word 'substituted'... as seen in the screenshot below]?? If so is it really worth to replace??
<img width="1082" alt="image" src="https://user-images.githubusercontent.com/51018572/200089578-144ec893-3a59-4e71-b30d-f1dcb54c8de3.png">

**Comment by ryanto** [comment]
Good catch! We'll a way to opt into other adapters (like xhr-http-handler) in this library. It might take me a little while to add that feature, but I think its the best way forward.

**Comment by nik32** [comment]
> Also does this completely replaces `fetch` as our http request handler [coz in the docs they use the word 'substituted'... as seen in the screenshot below]?? If so is it really worth it to replace for jus upload progress??
> <img alt="image" width="1082" src="https://user-images.githubusercontent.com/51018572/200089578-144ec893-3a59-4e71-b30d-f1dcb54c8de3.png">

@ryanto thank you for doing this!!! take your time in doing this, as we are not in a hurry for the next month or so.

**Comment by nik32** [comment]
@ryanto > Also does this completely replaces `fetch` as our http request handler [coz in the docs they use the word 'substituted'... as seen in the screenshot below]?? If so is it really worth it to replace for just upload progress??
> <img alt="image" width="1082" src="https://user-images.githubusercontent.com/51018572/200089578-144ec893-3a59-4e71-b30d-f1dcb54c8de3.png">

**Comment by chriship** [comment]
@ryanto Has any progress been made on this? I'm also noticing my upoads not reporting their progress until fully uploaded.
